File: tw/src/main/java/pl/edu/agh/philosophers/Philosopher.java
package pl.edu.agh.philosophers;

import pl.edu.agh.util.Utils;

import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

public class Philosopher extends Thread {
    private int number;
    private Fork fork;

    public Philosopher(int number, Fork fork) {
        this.number = number;
        this.fork = fork;
    }

    public void think() {

    }

    public void eat() {
        long waitTime = Utils.measureExecutionTime(() -> fork.take(number));
        Statistician.getInstance().addWaitTime(number, waitTime);
        fork.putBack(number);
    }

    @Override
    public void run() {
        for (int i = 0; i < 1000; i++) {
            think();
            eat();
        }
    }

    public static void main(String[] args) {
        Fork fork = new WaiterFork();
        List<Thread> philosophers = IntStream.range(0, 5)
                .mapToObj(i -> new Philosopher(i, fork)).collect(Collectors.toList());
        philosophers.forEach(Thread::start);
        philosophers.forEach(Utils::joinUnchecked);
        Utils.printMeasurements(Statistician.getInstance().waitTimeMean());
    }
}

File: philosopers_pyt/philosopher.py
import threading
import time

import statistician
from asymmetric_fork import AsymmetricFork
from blocking_fork import BlockingFork
from hunger_fork import HungerFork
from waiter_fork import WaiterFork


class Philosopher(threading.Thread):

    def __init__(self, number, fork):
        super().__init__()
        self.number = number
        self.fork = fork

    def think(self):
        pass

    def eat(self):
        start = time.clock()
        self.fork.take(self.number)
        end = time.clock()
        statistician.add_wait_time(self.number, end - start)
        self.fork.put_back(self.number)

    def run(self):
        for _ in range(1000):
            self.think()
            self.eat()


if __name__ == '__main__':
    fork = AsymmetricFork()
    philosophers = list(map(lambda i: Philosopher(i, fork), range(5)))
    for philosopher in philosophers:
        philosopher.start()
    for philosopher in philosophers:
        philosopher.join()
    statistician.print_measurements(statistician.wait_time_mean())
